cpp/dolfinx/common/sort.h

```diff
@@ -6,161 +6,105 @@
+  /// @tparam BITS The number of bits to sort at a time.
+  /// @param[in, out] range The range to sort.
+  /// @param[in] P Element projection.
+  template <std::ranges::random_access_range R, typename P = std::identity,
+            int BITS = 8>
+    requires std::integral<
+        std::remove_cvref_t<std::invoke_result_t<P, std::iter_value_t<R>>>>
+  constexpr void operator()(R&& range, P proj = {}) const
   {
-    // Zero counter array
-    std::ranges::fill(counter, 0);
-
-    // Count number of elements per bucket
-    for (T c : current_perm)
-      counter[(c & mask) >> mask_offset]++;
-
-    // Prefix sum to get the inserting position
-    offset[0] = 0;
-    std::partial_sum(counter.begin(), counter.end(), std::next(offset.begin()));
-    for (T c : current_perm)
-    {
-      std::int32_t bucket = (c & mask) >> mask_offset;
-      std::int32_t new_pos = offset[bucket + 1] - counter[bucket];
-      next_perm[new_pos] = c;
-      counter[bucket]--;
-    }
-
-    mask = mask << BITS;
-    mask_offset += BITS;
-
-    std::swap(current_perm, next_perm);
-  }
+    // value type
+    using T = std::iter_value_t<R>;
 
-  // Copy data back to array
-  if (its % 2 != 0)
-    std::ranges::copy(buffer, array.begin());
-}
-
-/// Returns the indices that would sort (lexicographic) a vector of
-/// bitsets.
-/// @tparam T The size of the bitset, which corresponds to the number of
-/// bits necessary to represent a set of integers. For example, N = 96
-/// for mapping three std::int32_t.
-/// @tparam BITS The number of bits to sort at a time
-/// @param[in] array The array to sort
-/// @param[in] perm FIXME
-template <typename T, int BITS = 16>
-void argsort_radix(std::span<const T> array, std::span<std::int32_t> perm)
-{
-  static_assert(std::is_integral_v<T>, "Integral required.");
-
-  if (array.size() <= 1)
-    return;
-
-  const auto [min, max] = std::ranges::minmax_element(array);
-  T range = *max - *min + 1;
+    // index type (if no projection is provided it holds I == T)
+    using I = std::remove_cvref_t<std::invoke_result_t<P, T>>;
 
-  // Sort N bits at a time
-  constexpr int bucket_size = 1 << BITS;
-  T mask = (T(1) << BITS) - 1;
-  std::int32_t mask_offset = 0;
+    if (range.size() <= 1)
+      return;
 
-  // Compute number of iterations, most significant digit (N bits) of
-  // maxvalue
-  int its = 0;
-  while (range)
-  {
-    range >>= BITS;
-    its++;
-  }
+    T max_value = proj(*std::ranges::max_element(range, std::less{}, proj));
 
-  // Adjacency list arrays for computing insertion position
-  std::array<std::int32_t, bucket_size> counter;
-  std::array<std::int32_t, bucket_size + 1> offset;
-
-  std::vector<std::int32_t> perm2(perm.size());
-  std::span<std::int32_t> current_perm = perm;
-  std::span<std::int32_t> next_perm = perm2;
-  for (int i = 0; i < its; i++)
-  {
-    // Zero counter
-    std::ranges::fill(counter, 0);
+    // Sort N bits at a time
+    constexpr int bucket_size = 1 << BITS;
```

Comment:
Is `int` the appropriate type?

Reply:
Implemented that. No more explicit int usage present.